src/plots/plots.js

```diff
@@ -563,6 +567,83 @@ function remapTransformedArrays(cd0, newTrace) {
     }
 }
 
+var formatKeys = [
+    'days', 'shortDays', 'months', 'shortMonths', 'periods',
+    'dateTime', 'date', 'time',
+    'decimal', 'thousands', 'grouping', 'currency'
+];
+
+/**
+ * getD3FormatObj: use _context to get the d3.locale argument object.
+ * decimal and thousands can be overridden later by layout.separators
+ * grouping and currency are not presently used by our automatic number
+ * formatting system but can be used by custom formats.
+ *
+ * @returns {object} d3.locale format object
+ */
+function getD3FormatObj(gd) {
+    var locale = gd._context.locale;
+    if(!locale) locale === 'en-US';
+
+    var formatDone = false;
+    var formatObj = {};
+
+    function includeFormat(newFormat) {
+        var formatFinished = true;
+        for(var i = 0; i < formatKeys.length; i++) {
+            var formatKey = formatKeys[i];
+            if(!formatObj[formatKey]) {
+                if(newFormat[formatKey]) {
+                    formatObj[formatKey] = newFormat[formatKey];
+                }
+                else formatFinished = false;
+            }
+        }
+        if(formatFinished) formatDone = true;
+    }
+
+    // same as localize, look for format parts in each format spec in the chain
+    for(var i = 0; i < 2; i++) {
+        var formats = gd._context.formats;
+        for(var j = 0; j < 2; j++) {
+            var formatj = formats[locale];
+            if(formatj) {
+                includeFormat(formatj);
+                if(formatDone) break;
+            }
+            formats = Registry.formatRegistry;
+        }
+
+        var baseLocale = locale.split('-')[0];
+        if(formatDone || baseLocale === locale) break;
+        locale = baseLocale;
+    }
+
+    // lastly pick out defaults from english (non-US, as DMY is so much more common)
+    if(!formatDone) includeFormat(Registry.formatRegistry.en);
+
+    return formatObj;
+}
+
+/**
+ * getFormatter: combine the final separators with the locale formatting object
+ * we pulled earlier to generate number and time formatters
+ * TODO: remove separators in v2, only use locale, so we don't need this step?
```

Comment:
I'm in strong agreement here too :ok_hand:

Reply:
Don't forget to add it #420 :smoking:

Reply:
That said, I just remembered that someone (in https://github.com/plotly/plotly.js/issues/1842) has asked for per-axis `separators` settings. I didn't really understand that person's use case though. Personally, I think `separators` are better served as locale/context options, but maybe I'm wrong.

Reply:
added https://github.com/plotly/plotly.js/issues/420#issuecomment-352099689

> That said, I just remembered that someone (in #1842) has asked for per-axis separators settings.

commented over there https://github.com/plotly/plotly.js/issues/1842#issuecomment-352105362